graylog2-server/src/main/java/org/graylog2/periodical/ContentPackLoaderPeriodical.java

```diff
@@ -0,0 +1,244 @@
+                                       @Named("content_packs_dir") Path contentPacksDir,
+                                       @Named("content_packs_auto_load") Set<String> contentPacksAutoLoad) {
+        this.objectMapper = objectMapper;
+        this.contentPackService = contentPackService;
+        this.contentPackPersistenceService = contentPackPersistenceService;
+        this.clusterConfigService = clusterConfigService;
+        this.userService = userService;
+        this.contentPacksLoaderEnabled = contentPacksLoaderEnabled;
+        this.contentPacksDir = contentPacksDir;
+        this.contentPacksAutoLoad = ImmutableSet.copyOf(contentPacksAutoLoad);
+    }
+
+    @Override
+    public boolean runsForever() {
+        return true;
+    }
+
+    @Override
+    public boolean stopOnGracefulShutdown() {
+        return false;
+    }
+
+    @Override
+    public boolean masterOnly() {
+        return true;
+    }
+
+    @Override
+    public boolean startOnThisNode() {
+        return contentPacksLoaderEnabled;
+    }
+
+    @Override
+    public boolean isDaemon() {
+        return true;
+    }
+
+    @Override
+    public int getInitialDelaySeconds() {
+        return 0;
+    }
+
+    @Override
+    public int getPeriodSeconds() {
+        return 0;
+    }
+
+    @Override
+    protected Logger getLogger() {
+        return LOG;
+    }
+
+    @Override
+    public void doRun() {
+        final ContentPackLoaderConfig contentPackLoaderConfig =
+                clusterConfigService.getOrDefault(ContentPackLoaderConfig.class, ContentPackLoaderConfig.EMPTY);
+
+        final List<Path> files = getFiles(contentPacksDir, FILENAME_GLOB);
+        final Map<String, ContentPack> contentPacks = new HashMap<>(files.size());
+
+        final Set<String> loadedContentPacks = new HashSet<>(contentPackLoaderConfig.loadedContentPacks());
+        final Set<String> appliedContentPacks = new HashSet<>(contentPackLoaderConfig.appliedContentPacks());
+        final Map<String, String> checksums = new HashMap<>(contentPackLoaderConfig.checksums());
+
+        for (Path file : files) {
+            final String fileName = file.getFileName().toString();
+
+            LOG.debug("Reading content pack from {}", file);
+            final byte[] bytes;
+            try {
+                bytes = Files.readAllBytes(file);
+            } catch (IOException e) {
+                LOG.warn("Couldn't read " + file + ". Skipping.", e);
+                continue;
+            }
+
+            final String encodedFileName = encodeFileNameForMongo(fileName);
+            final String checksum = HASH_FUNCTION.hashBytes(bytes).toString();
+            final String storedChecksum = checksums.get(encodedFileName);
+            if (storedChecksum == null) {
+                checksums.put(encodedFileName, checksum);
+            } else if (!checksum.equals(storedChecksum)) {
```

Comment:
I know that we used this checksum mechanism in the old content pack system, but I am not sure what it buys us. :smile: 

In the new content pack system every content pack has an ID and a revision. Wouldn't it be better to decide based on the ID and revision if we need to add a content pack to the database? If a content pack with the same ID and revision already exists, we skip the file.

The same goes for loaded (aka installed) content packs. Since the new system maintains installation state, we probably don't even need the `ContentPackLoaderConfig` cluster config object anymore, no?

What do you think?

Reply:
Sound right to me!